### ISSUE TITLE
MIDI issues on macOS

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#5622]
## Environment

* SuperCollider version: 3.12.1
* Operating system: macOS 10.15 (Catalina), macOS 11.6 (Big Sur)
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

```supercollider
MIDIClient.init;
m = MIDIOut.newByName("X-TOUCH COMPACT", "X-TOUCH COMPACT"); // testing with a midi cotroller
m.latency; // reports the default 0.2

m.noteOn(0, 30, 127); // this turns button LED on
// after the first start of SC after the boot, there's about 2s delay for the message to be sent (Catalina and Big Sur)
m.noteOn(0, 30, 0); // turn it off

// put computer to sleep for about a minute
m.noteOn(0, 30, 127); // now the message is delayed by over 40 seconds (the time the laptop was physically asleep) - Big Sur only
```

## TL;DR
As much as I'd like to keep the improved MIDI timing introduced in #4725, that PR seems to introduce issues that IMO are hard to be solved. Unless someone has an idea how to tackle them, I'd suggest to revert #4725 (and possibly release 3.12.2, if desired).

## Expected vs. actual behavior

#4725 improved the relative timing of MIDI messages, but introduced 2 (3) new issues:
1. all MIDI messages are delayed approx. 2s after the first run of `MIDIClient.init` after SC starts; re-running `MIDIClient.init` removes that delay. This happens on Catalina and on Big Sur
2. MIDI messages are additionally delayed by the time the computer was asleep since running `MIDIClient.init`. This happens on Big Sur only
3. If we use logical time for MIDI messages, latency calculation should probably be revised (right now the latency value is just added to the resulting logical timestamp). Depending how we deal with the above issues, this point might be moot if we revert #4725.

Details:
#4725 introduces definition of "starting time" using both SC thread time and time from `AudioHostGetTime()`. This seems like a good approach in principle. The issue is that starting with Big Sur, the time from `AudioHostGetTime()` does not advance when the computer is asleep, while the thread clock does advance. This conceptually similar to the issue solved in #5298 (also see explanation [here](https://github.com/supercollider/supercollider/issues/5168#issuecomment-749277265)) since that issue also dealt with comparing times from different references. Since MIDI message on macOS expects the time as reported from `AudioHostGetTime()`, I don't know how we could reliably calculate logical time, since the SC thread time does advance while the computer is asleep. I guess we could conceivably re-set that offset every time system wakes from sleep, but I'm not sure how to go about it or whether that's the best solution.

Regarding problem `1.`, there seems to be some fluctuation in thread timing that occurs only on the first MIDI initialization after SC starts. I don't know where that comes from.
I tried posting the offset between `AudioHostGetTime()` and SC thread when initializing MIDI and that offset is consistently off by approx. 2s between the first and the subsequent runs of `MIDIClient.init`. The subsequent runs report offsets that differ only by few ms, which is to be expected, I think, and is not an issue. Only that initial offset is considerably large. (See below for examining that delay)

I was trying to fix the issue with clock diverging when computer goes to sleep, but that didn't work since we ultimately need to use the time in relation to `AudioHostGetTime()` to schedule MIDI events. However, my code does post initial time difference between SC clock and system nanosecond clock, where one can observe the difference in offsets between the first and the subsequent initializations of MIDI client: https://github.com/dyfer/supercollider/tree/topic/macos-midi-01

**Comment by dyfer** [comment]
@telephon @sirlensalot if there are no more comments on this, I will propose a PR reverting the change from #4725. I find more and more cases where real-time use of MIDI becomes impossible because of this.

**Comment by telephon** [comment]
Yes, that makes sense. I can't judge the reasons that motivated #4725, so I don't know what we lose.

**Comment by dyfer** [comment]
The relative timing (i.e. rhythmic accuracy) did seem to improve after #4725. But I think we can't do that at the price of real-time usage of midi :)
I'll submit a PR.

**Comment by dyfer** [comment]
Fixed in #5631
